### ISSUE TITLE
User can register and log in

### ISSUE THREAD

**Issue by foohonpie** [ImperialConflict/imperialconflict.com#6]
## API

Docs in progress, but endpoints are functional.  Basic info is:

base url: `https://dev-api.imperialconflict.com`
registration endpoint: `POST /register`
auth endpoint: `POST /login`
auth user endpoint: `GET /user`

See "Fields and Validation" for input info.

## UI

No nav for now, just the register/login buttons on the home page that should trigger their respective modals.

## Fields and Validation

Registration Fields and FE validation are:
  * `name`
    * required
    * alphanum
  * `email_address`
    * required
    * email format
  * `password`
    * required
    * min:8
  * `confirm_password`
    * required
    * must match `password`

We should catch errors before sending to the API, which will do its own validation.

## Post-Register

Upon successful registration, a Bearer token will be received to be passed in subsequent calls.

The user should land on /gate.  The /gate page should send request with Bearer token to `GET /user`
   * If auth fails, send to home page with error
   * If auth succeeds, render simple dump of the return data

More detail will come for the "gate" page later on; this is just to get something functional for reg/auth.

Please post here with any questions.  Thanks!

**Comment by bpkennedy** [comment]
## UI and UX Questions
1. Is 'Creating An Empire' the same thing as Registration?
2. Are we foregoing the Reset Password functionality for the moment?
3. How should the modals look, style-wise?  Should I make them kinda look like the existing imperialconflict.com site or the pbbg.com site or something else (like default Quasar...)?  I am assuming they can be full-width/height on mobile screen sizes?
4. Should the user be able to get to the Registration modal *from* the Login modal and vice versa?  Kind of like a "Need an account? Sign up" link and a "Already have an account? Log in" link.
5. I will make the validation text a red color and put it above the buttons (as it is in IC login modal currently)?
6. Do you want me to setup that dynamic form/form field generation service like was done for pbbg.com?  Since it's done there it really wouldn't take much time to bring it in likewise here.
7. Do you want any Toast notifications (the little green-ish/red-ish colored panels that popup with text describing affirmative/negative)?  If so, where do you want them to be positioned as a kind of global default position?  And for which actions throughout the Registration/Login would you want to show one(s)?  And what would the text of them say?  I would think at minimum we can use the red/negative kind if login errors come back from the API (but the user's form submission was technically valid).

## Authentication
The first real potential vector for attacks/exploits/cheats!  So, we are establishing a Bearer token passed with every request (once user is logged in).  Great!  Did you want to make any adjustments from how this was setup for pbbg.com?  That setup is like: 
* token is sent in the Authorization header of every http request
* to facilitate browser reload and prevent needing to re-login we have to store the token somewhere client-side.  It is currently in LocalStorage.  The vulnerability here is that if a malicious player gets into another player's browser they could grab that token from the dev tools and put it into an http request of their own and essentially impersonate that user.  Now, they'd have to access the user's browser and the token would also have to still be valid/not expired.  My opinion is that this is unlikely - but I am not familiar with the lengths your playerbase may go to for game advantages. 😄 
* rather than allowing a given page to handle asking for `GET /user`, can I simply automatically fetch `/user` anytime that the user is logged in?  I think that will handle more use cases for you (i.e. like players link directly to pages other than the /gate in future).
* while the `/gate` page is an empty screen for the moment other than the `/user` data, what kind of high level layout are you wanting from quasar?  (You can play around with [the Layout Build here](https://quasar.dev/layout-builder) and export it when you're satisfied.  And I'm assuming that the very initial "home view" with the Create Empire/Log In buttons won't actually be in that layout, right?

**Comment by foohonpie** [comment]
Great questions!

> Is 'Creating An Empire' the same thing as Registration?

In this context, yes.  However there will be another "Join a Galaxy" process that will create the actual galaxy-specific empires (sub-accounts).  The phrasing of some of that may change at some point to be less confusing but for now we can just consider the user registration here.

> Are we foregoing the Reset Password functionality for the moment?

For now yeah, that'll come in another issue.

> How should the modals look, style-wise? Should I make them kinda look like the existing imperialconflict.com site or the pbbg.com site or something else (like default Quasar...)? I am assuming they can be full-width/height on mobile screen sizes?

The pbbg.com-style modals would work nicely here.

> Should the user be able to get to the Registration modal from the Login modal and vice versa? Kind of like a "Need an account? Sign up" link and a "Already have an account? Log in" link.

Yes, let's do that.

> I will make the validation text a red color and put it above the buttons (as it is in IC login modal currently)?

Sounds good!

> Do you want me to setup that dynamic form/form field generation service like was done for pbbg.com? Since it's done there it really wouldn't take much time to bring it in likewise here.

Yeah let's go for it; there will be lots of other forms coming into play that may benefit from that.

> Do you want any Toast notifications (the little green-ish/red-ish colored panels that popup with text describing affirmative/negative)? If so, where do you want them to be positioned as a kind of global default position? And for which actions throughout the Registration/Login would you want to show one(s)? And what would the text of them say? I would think at minimum we can use the red/negative kind if login errors come back from the API (but the user's form submission was technically valid).

Let's wait on this.  I see uses for them coming up, but not for reg/auth here.

> Did you want to make any adjustments from how this was setup for pbbg.com?

For now it's ok to use the same token approach as pbbg.com.

> rather than allowing a given page to handle asking for GET /user, can I simply automatically fetch /user anytime that the user is logged in? 

This sounds reasonable.  Good call.

> while the /gate page is an empty screen for the moment other than the /user data, what kind of high level layout are you wanting from quasar?

Ah yeah, I had intended to create a separate ticket for that.  I owe you some left-nav links, but let's go w this layout for authenticated user:

```
<template>
  <q-layout view="hHh LpR fFf">

    <q-header elevated class="bg-primary text-white">
      <q-toolbar>
        <q-btn dense flat round icon="menu" @click="left = !left" />

        <q-toolbar-title>
          <q-avatar>
            <img src="https://cdn.quasar.dev/logo/svg/quasar-logo.svg">
          </q-avatar>
          Title
        </q-toolbar-title>

        <q-btn dense flat round icon="menu" @click="right = !right" />
      </q-toolbar>
    </q-header>

    <q-drawer v-model="left" side="left" overlay bordered>
      <!-- drawer content -->
    </q-drawer>

    <q-drawer v-model="right" side="right" overlay bordered>
      <!-- drawer content -->
    </q-drawer>

    <q-page-container>
      <router-view />
    </q-page-container>

  </q-layout>
</template>

<script>
export default {
  data () {
    return {
      left: false,
      right: false
    }
  }
}
</script>
```

Thanks!
